### PR TITLE
WideResNet

### PR DESCRIPTION
**Pull request by ferdinandlouapre**
Add growing dropout to resnet container

### CODE REVIEW

File: src/gromo/containers/resnet.py [hunk 1]
```diff
@@ -843,7 +856,8 @@ def init_full_resnet_structure(
     print(f"Number of parameters (classical): {classical_params}")
 
     # Compare with torchvision ResNet-18
-    import torchvision.models as models  # pyright: ignore[reportMissingImports]
+    # pyright: ignore[reportMissingImports]
+    import torchvision.models as models
```
Comment (sylvchev) [inline src/gromo/containers/resnet.py:860]: If you put `# pyright` at the end of the line, it is a special instruction for one line, but if you put it a the beginning, it is a whole file instructions.
If the line is too long you could drop the `[reportMissingImports]` and put a comment beforehands.

File: src/gromo/modules/growing_dropout.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+import torch
+import torch.nn as nn
+
+
+class GrowingDropout(nn.modules.dropout._DropoutNd):
+    """
+    Base class for dropout layers on growing architectures.
+
+    This class provides the common functionality for dropout
+    layers on basic architecture without penalyzing growth.
```
Comment (sylvchev) [inline src/gromo/modules/growing_dropout.py]:
```suggestion
    This class provides dropout without penalyzing growth.
```

File: src/gromo/modules/growing_dropout.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    The growth of the previous layer is never zeroed.
+
+    Parameters
+    ----------
+    dropout_rate : float
+        Probability of an element to be zeroed.
+        Set to non 0 to run WideResNet architecture.
+        Default: 0.0 for classical ResNet.
```
Comment (sylvchev) [inline src/gromo/modules/growing_dropout.py]:
```suggestion
        Set to non-zero to activate.
        Default: 0.0 for no dropout.
```

**Review by TheoRudkiewicz — CHANGES_REQUESTED** [review]
Great ! If you could add a  minimal test showing how to instantiate classical `torchvision` model and showing it has the same number of parameters as the one of the library. (In a similar way has it's done in the `test_vgg`).

**Review by sylvchev — COMMENTED** [review]
Thanks @ferdinandlouapre 
Did you try the code on small experiments and verify if this is inline with reference (non-growing) implementation ? If so, could you share some plots/tables here so we could keep them here for later use?
Overall, this looks nice and I could merge the PR when the changes are applied.

**Comment by ferdinandlouapre** [comment]
I checkout on my personal repo, I have the same number of parameters as WideResnet28-1 imported from https://github.com/bmsookim/wide-resnet.pytorch.git *(with bias enable for every convolution)*

**Review by TheoRudkiewicz — APPROVED** [review]
Good for me! Thanks @ferdinandlouapre

**Comment by ferdinandlouapre** [comment]
@sylvchev Voici quelques résultats que j'ai obtenu en entrainant 3 modèles WideResNet 28-1 avec dropout 30% sur CIFAR10 avec un reducing factor de 64

| Name | num_parameters_init | num_parameters_end | final_val_accuracy | final_test_accuracy | final_train_accuracy | final_val_loss | final_test_loss | final_train_loss |
| :--- | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| fogro | 27474.0 | 392676.0 | 0.907 | 0.905 | 0.897 | 0.276 | 0.286 | 0.297 |
| full_resnet | 369594.0 | 369594.0 | 0.917 | 0.910 | 0.909 | 0.262 | 0.272 | 0.261 |
| random_zeros | 27474.0 | 363794.0 | 0.903 | 0.901 | 0.895 | 0.295 | 0.302 | 0.304 |

**Comment by ferdinandlouapre** [comment]
image
image
image

**Comment by sylvchev** [comment]
Good for me, we could merge. The architecture is 5 points below the reference value, but that are ultra-optimized (lr adapted by hand every 50 epochs 🙄 ).
If you have the test accuracy could you add it here as well?

**Review by sylvchev — APPROVED** [review]

**Comment by ferdinandlouapre** [comment]
> Good for me, we could merge. The architecture is 5 points below the reference value, but that are ultra-optimized (lr adapted by hand every 50 epochs 🙄 ). If you have the test accuracy could you add it here as well?

I updated the previous table with test accuracy and loss.
